### PR TITLE
[FIXED JENKINS-33201 Java JSON exception with an empty parametrized build.]

### PR DESCRIPTION
**Pull request by rupinr**
I have handled this issue gracefully. Now user will be redirected to an error page giving information about the error (Similar to how it is handled for empty job/view name)

image

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -4826,5 +4826,20 @@ private static void _setJenkinsJVM(boolean jenkinsJVM) {
...
+     * but no parameters are selected. User will be redirected to an error page
+     * with proper error message.
+     * @param jsonProperties
+     */
+    public static void checkForEmptyParameters(JSONObject jsonProperties){
+    	JSONObject parameterDefinitionProperty = jsonProperties
+				.getJSONObject("hudson-model-ParametersDefinitionProperty");
+		if ((parameterDefinitionProperty.getBoolean("specified") == true)
+				&& !parameterDefinitionProperty.has("parameterDefinitions")) {
+			throw new Failure(Messages.Hudson_NoParamsSpecified());
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: Should throw FormException at least

**Comment by oleg-nenashev** [comment]
I think it would be better to just ignore empty property instead of throwing exxceptions

**Comment by rupinr** [comment]
I have changed to FormException. Please check now.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -4826,5 +4826,21 @@ private static void _setJenkinsJVM(boolean jenkinsJVM) {
...
+    /**
+     * This handles the situation when Parameterized build checkbox is checked 
+     * but no parameters are selected. User will be redirected to an error page
+     * with proper error message.
+     * @param jsonProperties
+     * @throws FormException 
+     */
+    public static void checkForEmptyParameters(JSONObject jsonProperties) throws FormException{
+    	JSONObject parameterDefinitionProperty = jsonProperties
+				.getJSONObject("hudson-model-ParametersDefinitionProperty");
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: This file uses tabs, not spaces

File: core/src/main/resources/hudson/model/Messages.properties [hunk 3]
```diff
@@ -385,3 +385,5 @@ Jenkins.IsRestarting=Jenkins is restarting
 
 User.IllegalUsername="{0}" is prohibited as a username for security reasons.
 User.IllegalFullname="{0}" is prohibited as a full name for security reasons.
+
+Hudson.NoParamsSpecified=No Parameters are specified for this parameterized build
```
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/model/Messages.properties]: NIT: no line break at the end

**Comment by oleg-nenashev** [comment]
Seems you've rebased some master branch changes instead of merging them. Not a problem if we squash the PR during merging.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -4826,5 +4826,18 @@ private static void _setJenkinsJVM(boolean jenkinsJVM) {
...
         }
     }
+    /**
+     * This handles the situation when Parameterized build checkbox is checked 
+     * but no parameters are selected. User will be redirected to an error page
+     * with proper error message.
+     * @param jsonProperties
+     * @throws FormException 
+     */
+    public static void checkForEmptyParameters(JSONObject jsonProperties) throws FormException{
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: I do not see this method as a part of public API (especially in the Jenkins class). There are hardcoded values for jobs only. Would be better to just inline this method to the code or to move it to a private method in the Job class

Comment (rupinr) [inline core/src/main/java/jenkins/model/Jenkins.java]: Done.

**Comment by rupinr** [comment]
And regarding the comment "

> Seems you've rebased some master branch changes instead of merging them. Not a problem if we squash the PR during merging.

Please squash these commits, so that my commits wont mess up the git log in master  :+1:

File: core/src/main/java/hudson/model/Job.java [hunk 5]
```diff
@@ -1509,4 +1511,18 @@ public BuildTimelineWidget getTimeline() {
     }
 
     private final static HexStringConfidentialKey SERVER_COOKIE = new HexStringConfidentialKey(Job.class,"serverCookie",16);
+    
+    /**
+     * This handles the situation when Parameterized build checkbox is checked 
+     * but no parameters are selected. User will be redirected to an error page
+     * with proper error message.
+     * @param jsonProperties
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Job.java]: Just for the future. Please avoid using empty Javadoc definitions. Default Javadoc linting behavior in JDK8 considers it as an error

**Comment by oleg-nenashev** [comment]
weak 👍 for merging with squash

**Comment by rupinr** [comment]
> weak :+1: for merging with squash
Do I need to resubmit my PR with a clean history?..

> Just for the future. Please avoid using empty Javadoc definitions. Default Javadoc linting behavior in JDK8 considers it as an error

I will take care of it in future.

**Comment by Vlatombe** [comment]
@rupinr this is instruction for the merger (GitHub provides directly the ability to squash before merge)

**Comment by rupinr** [comment]
@Vlatombe So, I guess nothing is pending from my side. Right?

**Comment by oleg-nenashev** [comment]
@rupinr Merged the PR. Thanks for your contribution!

**Comment by serafeimgr** [comment]
Thanks rupinr.
I am the reporter of this issue, should i close it?

File: core/src/main/java/hudson/model/Job.java [hunk 6]
```diff
@@ -1509,4 +1511,18 @@ public BuildTimelineWidget getTimeline() {
...
+     * @throws FormException 
+     */
+    private static void checkForEmptyParameters(JSONObject jsonProperties) throws FormException{
+        JSONObject parameterDefinitionProperty = jsonProperties.getJSONObject("hudson-model-ParametersDefinitionProperty");
+        if ((parameterDefinitionProperty.getBoolean("specified") == true)&& !parameterDefinitionProperty.has("parameterDefinitions")) {
+		    throw new FormException(Messages.Hudson_NoParamsSpecified(),"parameterDefinitions");
```
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java]: As noted in #2589, I think this is overkill. Better to just ignore the checkbox.

**Pull request merged**